File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        if(root == NULL)
            return 0;
        queue<pair<TreeNode*, long long>> q;
        long long maxCount=1;
        q.push({root, 0}); 
        while(q.size()>0){
            int size=q.size();
            long long start = q.front().second;
            long long end = q.back().second;
            maxCount = max(maxCount,end-start + 1);
            for(int i = 0 ; i < size ; i++){
                pair<TreeNode*, int> p = q.front();
                long long idx = p.second - start;
                q.pop();
                
                if(p.first->left != NULL)
                    q.push({p.first->left, (long long)2 * idx + 1});
                
                if(p.first->right != NULL)
                    q.push({p.first->right, (long long) 2 * idx + 2});
            }
        }
        return maxCount;
    }
};

File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public int widthOfBinaryTree(TreeNode root) {
        if (root == null) {
            return 0;
        }
        Queue<Pair<TreeNode, Long>> q = new LinkedList<>();
        long res = 1;
        q.add(new Pair<>(root, 0L));
        while (!q.isEmpty()) {
            long n = q.size();
            long s = q.peek().getValue();
            long e = 0;
            for (int i = 0; i < n; i++) {
                e = q.peek().getValue();
                if (q.peek().getKey().left != null) {
                    q.add(new Pair<>(q.peek().getKey().left, 2 * e + 1));
                }
                if (q.peek().getKey().right != null) {
                    q.add(new Pair<>(q.peek().getKey().right, 2 * e + 2));
                }
                q.poll();
            }
            res = Math.max(res, e - s + 1);
        }
        return (int) res;
    }
}